src/lib/transition_frontier_persistence/transition_frontier_persistence.ml

```diff
@@ -49,57 +66,84 @@ module Make (Inputs : Intf.Main_inputs) = struct
             (Transition_frontier.Diff_hash.to_string new_hash)
             ()
 
-  let create ?directory_name ~logger max_buffer_capacity =
+  let flush {buffer; worker_mvar; _} =
+    let list = Queue.to_list buffer in
+    Queue.clear buffer ;
+    Mvar.put worker_mvar (`Continue list)
+
+  let create ?directory_name ~logger ~flush_capacity ~max_buffer_capacity () =
     let worker = Worker.create ?directory_name ~logger () in
-    let reader, writer = Pipe.create () in
     let buffer = Queue.create () in
+    let worker_mvar = Mvar.create () in
+    let stop_signal = Ivar.create () in
     let worker_thread =
-      Pipe.fold ~flushed:Pipe.Flushed.When_value_processed
-        ~init:Transition_frontier.Diff_hash.empty reader
-        ~f:(fun init_hash diff_pairs ->
-          Deferred.List.fold diff_pairs ~init:init_hash
-            ~f:(fun acc_hash
-               (Transition_frontier.Diff_mutant.E.Pair
-                 (diff, ground_truth_mutant))
-               ->
-              let%bind result =
-                write_diff_and_verify ~logger ~acc_hash worker
-                  (diff, ground_truth_mutant)
-              in
-              (* We would want the scheduler to run other jobs after computing a diff *)
-              Deferred.create @@ fun ivar -> Ivar.fill ivar result ) )
-      |> Deferred.ignore
-    in
-    {worker; reader; writer; max_buffer_capacity; buffer; worker_thread}
-
-  let close {worker; writer; _} = Pipe.close writer ; Worker.close worker
+      let rec process_diff_work init_hash =
+        match%bind Mvar.take worker_mvar with
+        | `Finished ->
+            Deferred.unit
+        | `Continue diff_pairs ->
+            let%bind new_hash =
+              Deferred.List.fold diff_pairs ~init:init_hash
+                ~f:(fun acc_hash
+                   (Transition_frontier.Diff_mutant.E.With_value
+                     (diff, ground_truth_mutant))
+                   ->
+                  write_diff_and_verify ~logger ~acc_hash worker
+                    (diff, ground_truth_mutant) )
+            in
+            process_diff_work new_hash
+      in
+      process_diff_work Transition_frontier.Diff_hash.empty
+    in
+    let t =
+      { worker
+      ; worker_mvar= Mvar.write_only worker_mvar
+      ; flush_capacity
+      ; max_buffer_capacity
+      ; buffer
+      ; stop_signal
+      ; worker_thread }
+    in
+    let stop_thread =
+      let%bind `Stop = Ivar.read stop_signal in
+      let%bind () = flush t in
+      Mvar.put worker_mvar `Finished
+    in
+    don't_wait_for stop_thread ; t
 
-  let flush {writer; buffer; _} =
-    let list = Queue.to_list buffer in
-    Pipe.write_without_pushback writer list ;
-    Queue.clear buffer
+  let close {worker; _} = Worker.close worker
```

Comment:
Isn't this function unsafe to call by itself? Probably you should only have `close_and_finish_copy`